src/zope/password/password.py

```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
+    slower each attempt will be.
+
+    Compared to the :class:`BCRYPTPasswordManager`, this has the
+    advantage of allowing tunable rounds, so as computing devices get
+    more powerful making brute force attacks faster, the difficulty
+    level can be raised (for newly encoded passwords).
+
+    >>> from zope.password.password import BCRYPTKDFPasswordManager
+    >>> manager = BCRYPTKDFPasswordManager()
+    >>> manager.checkPassword(b'not from here', None)
+    False
+
+    Let's encode a password. We'll use the minimum acceptable number
+    of rounds so that the tests run fast:
+
+    >>> manager.rounds = 51
+    >>> password = u"right \N{CYRILLIC CAPITAL LETTER A}"
+    >>> encoded = manager.encodePassword(password)
+    >>> print(encoded.decode())
+    {BCRYPTKDF}33...
+
+    It checks out:
+
+    >>> manager.checkPassword(encoded, password)
+    True
+
+    We can change the number of rounds for future encodings:
+
+    >>> manager.rounds = 100
+    >>> encoded2 = manager.encodePassword(password)
+    >>> print(encoded2.decode())
+    {BCRYPTKDF}64...
+    >>> manager.checkPassword(encoded2, password)
+    True
+
+    And the old password still checks out:
+
+    >>> manager.checkPassword(encoded, password)
+    True
+    """
+
+    #: The number of rounds of hashing that should be applied.
+    #: The higher the number, the slower it is. It should be at least
+    #: 50.
+    rounds = 1<<10
+
+    #: The number of bytes long the encoded password will be. It must be
+    #: at least 1 and no more than 512.
+    keylen = 32
+
+    _prefix = b'{BCRYPTKDF}'
+
+    def _encode(self, password, salt, rounds, keylen):
+        password = _encoder(password)
+
+        key = bcrypt.kdf(password, salt=salt,
+                         desired_key_bytes=keylen,
+                         rounds=rounds)
+        rounds_bytes = _encoder('%x' % rounds)
+        result = (self._prefix
+                  + rounds_bytes
+                  + b'$'
+                  + urlsafe_b64encode(salt)
+                  + b'$'
+                  + urlsafe_b64encode(key))
+        return result
+
+    def encodePassword(self, password):
+        salt = bcrypt.gensalt()
+        return self._encode(password, salt, self.rounds, self.keylen)
+
+    def checkPassword(self, hashed_password, clear_password):
+        hashed_password = _encoder(hashed_password)
+        if not self.match(hashed_password):
+            return False
+        rounds, salt, key = hashed_password[len(self._prefix):].split(b'$')
+        rounds = int(rounds, 16)
+
+        salt = urlsafe_b64decode(salt)
+        keylen = len(urlsafe_b64decode(key))
+        encoded_password = self._encode(clear_password, salt, rounds, keylen)
+        # XXX: Do we need to use a timing-safe comparison function here?
+        # z3c.bcrypt does, and so does bcrypt.hashpw. But there is no such
+        # public function
+        return hashed_password == encoded_password
```

Comment:
[hmac.compare_digest()](https://docs.python.org/3/library/hmac.html#hmac.compare_digest) is such a function, but only available in Python starting with 3.3.

Reply:
Thanks, I'll add that to the comment.

Reply:
Actually, I'll just use it when available. Thanks.

Reply:
Interestingly, `hmac.compare_digest` was actually [backported to 2.7.7](https://github.com/python/cpython/commit/629026aecc6bbb7cf8df25f7699eaf97d8e45701)  in 2014.  Do we need to care about supporting older point releases in a new version? I can add appropriate `requires-python` metadata to make sure a new release is not installed by older versions.

Reply:
> Do we need to care about supporting older point releases in a new version?

Given that this is a security feature for a security-critical package, and that people who care about security are probably running 2.7.9+ anyway, I feel like it is going to be OK to require 2.7.7+ (a three-year-old release) in a new version. Especially since it adds new security-related features. And especially especially since the requires-python metadata generally means it won't get installed by accident. 

I added this to the PR.